File: server/routes/masters.js
import express from 'express';
import { db, saveDb } from '../db/database.js';

const router = express.Router();

// ITEMS API
router.get('/items', (req, res) => {
  const { search, category, warehouse } = req.query;
  let items = [...db.items];

  if (category) items = items.filter(i => i.category_id === category);
  if (search) {
    const q = search.toLowerCase();
    items = items.filter(i => 
      i.item_code.toLowerCase().includes(q) ||
      i.item_name.toLowerCase().includes(q) ||
      (i.description && i.description.toLowerCase().includes(q))
    );
  }

  // Populate references & stock balances
  const result = items.map(item => {
    const cat = db.item_categories.find(c => c.id === item.category_id) || {};
    const brand = db.brands.find(b => b.id === item.brand_id) || {};
    const uom = db.units_of_measure.find(u => u.id === item.uom_id) || {};
    const tax = db.tax_rates.find(t => t.id === item.tax_rate_id) || {};
    const balances = db.inventory_balances.filter(b => b.item_id === item.id);
    
    const onHand = balances.reduce((sum, b) => sum + (b.on_hand_qty || 0), 0);
    const reserved = balances.reduce((sum, b) => sum + (b.reserved_qty || 0), 0);
    const available = onHand - reserved;

    return {
      ...item,
      category_name: cat.category_name,
      brand_name: brand.brand_name,
      uom_symbol: uom.unit_symbol,
      tax_percentage: tax.percentage,
      on_hand_qty: onHand,
      reserved_qty: reserved,
      available_qty: available,
      stock_value: onHand * item.valuation_rate
    };
  });

  res.json({ success: true, data: result });
});

router.post('/items', (req, res) => {
  const newItem = {
    id: `itm-${Date.now()}`,
    item_code: req.body.item_code || `ITM-${Date.now()}`,
    item_name: req.body.item_name,
    description: req.body.description || '',
    category_id: req.body.category_id,
    brand_id: req.body.brand_id,
    uom_id: req.body.uom_id,
    purchase_uom_id: req.body.purchase_uom_id || req.body.uom_id,
    tax_rate_id: req.body.tax_rate_id,
    hsn_sac_code: req.body.hsn_sac_code || '',
    min_stock_level: Number(req.body.min_stock_level || 0),
    max_stock_level: Number(req.body.max_stock_level || 100),
    reorder_level: Number(req.body.reorder_level || 10),
    reorder_qty: Number(req.body.reorder_qty || 20),
    valuation_rate: Number(req.body.valuation_rate || 0),
    default_location_id: req.body.default_location_id || 'loc-01',
    is_batch_tracked: Boolean(req.body.is_batch_tracked),
    is_serial_tracked: Boolean(req.body.is_serial_tracked),
    is_expiry_tracked: Boolean(req.body.is_expiry_tracked),
    barcode: req.body.barcode || `${Math.floor(1000000000000 + Math.random() * 9000000000000)}`,
    image_url: req.body.image_url || 'https://images.unsplash.com/photo-1586075010923-2dd4570fb338?w=300&auto=format&fit=crop&q=60',
    is_active: true,
    created_at: new Date().toISOString()
  };

  db.items.push(newItem);
  
  // Initial zero balance
  db.inventory_balances.push({
    id: `bal-${Date.now()}`,
    item_id: newItem.id,
    warehouse_id: 'wh-01',
    location_id: newItem.default_location_id,
    on_hand_qty: 0,
    reserved_qty: 0,
    available_qty: 0,
    valuation_rate: newItem.valuation_rate
  });

  saveDb();
  res.json({ success: true, data: newItem, message: 'Item created successfully' });
});

// SUPPLIERS API
router.get('/suppliers', (req, res) => {
  res.json({ success: true, data: db.suppliers });
});

router.post('/suppliers', (req, res) => {
  const newSup = {
    id: `sup-${Date.now()}`,
    supplier_code: req.body.supplier_code || `SUP-${Math.floor(10000 + Math.random() * 90000)}`,
    supplier_name: req.body.supplier_name,
    contact_person: req.body.contact_person,
    phone: req.body.phone,
    email: req.body.email,
    address_registered: req.body.address_registered || '',
    address_billing: req.body.address_billing || '',
    gst_number: req.body.gst_number || '',
    pan_number: req.body.pan_number || '',
    payment_terms: req.body.payment_terms || 'Net 30 days',
    delivery_lead_time_days: Number(req.body.delivery_lead_time_days || 5),
    rating: 5.0,
    approval_status: 'Approved',
    bank_details: req.body.bank_details || {},
    is_active: true
  };
  db.suppliers.push(newSup);
  saveDb();
  res.json({ success: true, data: newSup, message: 'Supplier created successfully' });
});

// DEPARTMENTS API
router.get('/departments', (req, res) => {
  res.json({ success: true, data: db.departments });
});

// WAREHOUSES API
router.get('/warehouses', (req, res) => {
  const data = db.warehouses.map(w => {
    const locations = db.warehouse_locations.filter(l => l.warehouse_id === w.id);
    return { ...w, locations };
  });
  res.json({ success: true, data });
});

// CATEGORIES, BRANDS, UOMS, TAXES
router.get('/categories', (req, res) => res.json({ success: true, data: db.item_categories }));
router.get('/brands', (req, res) => res.json({ success: true, data: db.brands }));
router.get('/uoms', (req, res) => res.json({ success: true, data: db.units_of_measure }));
router.get('/taxes', (req, res) => res.json({ success: true, data: db.tax_rates }));

export default router;

File: server/routers/masters.py
from fastapi import APIRouter, HTTPException
from server.db.database_store import db, save_db
from server.schemas.schemas import ItemCreate, SupplierCreate
import uuid
from datetime import datetime

router = APIRouter(prefix="/api", tags=["Masters"])

# Items Master
@router.get("/items")
def get_items():
    return {"success": True, "data": db["items"]}

@router.post("/items")
def create_item(item: ItemCreate):
    new_id = f"itm-{str(len(db['items']) + 1).zfill(2)}"
    item_code = item.item_code or f"ITM-{datetime.now().strftime('%Y%m%d')}-{len(db['items']) + 1}"
    
    new_item = {
        "id": new_id,
        "item_code": item_code,
        "item_name": item.item_name,
        "description": item.description,
        "category_id": item.category_id or "cat-01",
        "brand_id": item.brand_id or "brd-01",
        "uom_id": item.uom_id or "uom-01",
        "purchase_uom_id": item.purchase_uom_id or "uom-01",
        "tax_rate_id": item.tax_rate_id or "tax-18",
        "hsn_sac_code": item.hsn_sac_code or "84713010",
        "min_stock_level": item.min_stock_level or 5,
        "max_stock_level": item.max_stock_level or 50,
        "reorder_level": item.reorder_level or 10,
        "reorder_qty": item.reorder_qty or 15,
        "valuation_rate": item.valuation_rate or 1000,
        "default_location_id": item.default_location_id or "loc-01",
        "is_batch_tracked": item.is_batch_tracked or False,
        "is_serial_tracked": item.is_serial_tracked or False,
        "is_expiry_tracked": item.is_expiry_tracked or False,
        "barcode": item.barcode or f"890123456789{len(db['items']) + 1}",
        "image_url": item.image_url or "https://images.unsplash.com/photo-1588872657578-7efd1f1555ed?w=300&auto=format&fit=crop&q=60",
        "is_active": True,
        "created_at": datetime.now().isoformat()
    }
    
    db["items"].append(new_item)

    # Initialize balance record
    balance_record = {
        "id": f"bal-{len(db['inventory_balances']) + 1}",
        "item_id": new_id,
        "warehouse_id": "wh-01",
        "location_id": "loc-01",
        "on_hand_qty": 20,
        "reserved_qty": 0,
        "available_qty": 20,
        "valuation_rate": item.valuation_rate or 1000
    }
    db["inventory_balances"].append(balance_record)
    save_db()

    return {"success": True, "data": new_item}

# Categories, Brands, UOM, Tax Rates
@router.get("/item-categories")
def get_categories():
    return {"success": True, "data": db["item_categories"]}

@router.get("/brands")
def get_brands():
    return {"success": True, "data": db["brands"]}

@router.get("/uom")
def get_uom():
    return {"success": True, "data": db["units_of_measure"]}

@router.get("/tax-rates")
def get_tax_rates():
    return {"success": True, "data": db["tax_rates"]}

# Suppliers Master
@router.get("/suppliers")
def get_suppliers():
    return {"success": True, "data": db["suppliers"]}

@router.post("/suppliers")
def create_supplier(sup: SupplierCreate):
    new_id = f"sup-{str(len(db['suppliers']) + 1).zfill(2)}"
    sup_code = sup.supplier_code or f"SUP-{str(len(db['suppliers']) + 48).zfill(5)}"
    
    new_supplier = {
        "id": new_id,
        "supplier_code": sup_code,
        "supplier_name": sup.supplier_name,
        "contact_person": sup.contact_person,
        "phone": sup.phone,
        "email": sup.email,
        "address_registered": sup.address_registered,
        "gst_number": sup.gst_number,
        "pan_number": sup.pan_number,
        "payment_terms": sup.payment_terms,
        "delivery_lead_time_days": sup.delivery_lead_time_days,
        "rating": 4.5,
        "approval_status": "Approved",
        "is_active": True
    }
    db["suppliers"].append(new_supplier)
    save_db()
    return {"success": True, "data": new_supplier}

# Departments & Warehouses
@router.get("/departments")
def get_departments():
    return {"success": True, "data": db["departments"]}

@router.get("/warehouses")
def get_warehouses():
    return {"success": True, "data": db["warehouses"]}

@router.get("/warehouse-locations")
def get_warehouse_locations():
    return {"success": True, "data": db["warehouse_locations"]}
